hyperspace/cosmos/src/provider.rs

```diff
@@ -0,0 +1,1149 @@
+				.await
+				.map_err(|e| Error::from(format!("Web Socket Client Error {:?}", e)))
+				.unwrap();
+			subscriptions.push(subscription);
+		}
+		// Collect IBC events from each RPC event, Like what `stream_batches()` does in `hermes`
+		let all_subs: Box<dyn Stream<Item = Result<Event, RpcError>> + Send + Sync + Unpin> =
+			Box::new(select_all(subscriptions));
+		let chain_id = self.chain_id.clone();
+		let events = all_subs
+			.map(move |event| {
+				// Like what `get_all_events()` does in `hermes`
+				let mut events_with_height: Vec<IbcEventWithHeight> = vec![];
+				let Event { data, events: _, query } = event.unwrap();
+				match data {
+					EventData::NewBlock { block, .. }
+						if query == Query::from(EventType::NewBlock).to_string() =>
+					{
+						let height = Height::new(
+							ChainId::chain_version(chain_id.to_string().as_str()),
+							u64::from(block.as_ref().ok_or("tx.height").unwrap().header.height),
+						);
+						events_with_height.push(IbcEventWithHeight::new(
+							ClientEvents::NewBlock::new(height).into(),
+							height,
+						));
+					},
+					EventData::Tx { tx_result } => {
+						let height = Height::new(
+							ChainId::chain_version(chain_id.to_string().as_str()),
+							tx_result.height as u64,
+						);
+						for abci_event in &tx_result.result.events {
+							if let Ok(ibc_event) = ibc_event_try_from_abci_event(abci_event, height)
+							{
+								if query == Query::eq("message.module", "ibc_client").to_string() &&
+									event_is_type_client(&ibc_event)
+								{
+									events_with_height
+										.push(IbcEventWithHeight::new(ibc_event, height));
+								} else if (query ==
+									Query::eq("message.module", "ibc_connection").to_string() ||
+									query ==
+										Query::eq("message.module", "ibc_client").to_string()) &&
+									event_is_type_connection(&ibc_event)
+								{
+									events_with_height
+										.push(IbcEventWithHeight::new(ibc_event, height));
+								} else if query ==
+									Query::eq("message.module", "ibc_channel").to_string() &&
+									event_is_type_channel(&ibc_event)
+								{
+									events_with_height
+										.push(IbcEventWithHeight::new(ibc_event, height));
+								}
+							} else {
+								log::debug!(target: "hyperspace_cosmos", "Failed to parse event {:?}", abci_event);
+							}
+						}
+					},
+					_ => {},
+				}
+				stream::iter(events_with_height)
+			})
+			.flatten()
+			.map(|e| e.event)
+			.boxed();
+		events
+	}
+
+	async fn query_client_consensus(
+		&self,
+		at: Height,
+		client_id: ClientId,
+		consensus_height: Height,
+	) -> Result<QueryConsensusStateResponse, Self::Error> {
+		let path_bytes = Path::ClientConsensusState(ClientConsensusStatePath {
+			client_id: client_id.clone(),
+			epoch: consensus_height.revision_number,
+			height: consensus_height.revision_height,
+		})
+		.to_string()
+		.into_bytes();
+		let (query_result, proof) = self.query_path(path_bytes.clone(), at, true).await?;
```

Comment:
IIRC, before, while I was implementing this section, `query_path` calls were not returning the proofs, which made me use `query_proof` method instead. Maybe you test it. However, having Tendermint-rs deps updated, the `query_path` method should be fine now. :ok_hand: